Implement `MaybeUninit`

japaric opened the pull request:
This PR:

- Adds `MaybeUninit` (see #53491) to `{core,std}::mem`.
- Makes `mem::{uninitialized,zeroed}` panic when they are used to instantiate an uninhabited type.
- Does *not* deprecate `mem::{uninitialized,zeroed}` just yet. As per https://github.com/rust-lang/rust/issues/53491#issuecomment-414147666, we should not deprecate them until `MaybeUninit` is stabilized.
- It replaces uses of `mem::{uninitialized,zeroed}` in core and alloc with `MaybeUninit`.

There are still several instances of `mem::{uninitialized,zeroed}` in `std` that *this* PR doesn't address.

r? @RalfJung 
cc @eddyb you may want to look at the new panicking logic

Mark-Simulacrum commented on src/libcore/ptr.rs:
```diff
@@ -142,16 +142,12 @@ pub const fn null_mut<T>() -> *mut T { 0 as *mut T }
 
     // Perform the swap
-    copy_nonoverlapping(x, &mut tmp, 1);
+    copy_nonoverlapping(x, tmp.get_mut(), 1);
```
Shouldn't this be using `as_mut_ptr`?

Mark-Simulacrum commented on src/libcore/ptr.rs:
```diff
@@ -296,9 +292,9 @@ pub unsafe fn replace<T>(dest: *mut T, mut src: T) -> T {
-    copy_nonoverlapping(src, &mut tmp, 1);
-    tmp
+    let mut tmp = MaybeUninit::<T>::uninitialized();
+    copy_nonoverlapping(src, tmp.get_mut(), 1);
```
This also looks like it should be `as_mut_ptr`? `tmp` is *not* initialized at this point.

Mark-Simulacrum commented on src/libcore/slice/rotate.rs:
```diff
@@ -88,8 +85,8 @@ pub unsafe fn ptr_rotate<T>(mut left: usize, mid: *mut T, mut right: usize) {
-    let rawarray = RawArray::new();
-    let buf = rawarray.ptr();
+    let rawarray = MaybeUninit::<RawArray<T>>::uninitialized();
+    let buf = rawarray.get_ref().ptr();
```
Isn't this also not legitimate code? `get_ref` is only defined as far as I can tell on initialized memory, but it's not initialized here.

Mark-Simulacrum commented on src/libcore/slice/sort.rs:
```diff
@@ -272,8 +272,8 @@ fn partition_in_blocks<T, F>(v: &mut [T], pivot: &T, is_less: &mut F) -> usize
-            start_l = offsets_l.as_mut_ptr();
-            end_l = offsets_l.as_mut_ptr();
+            start_l = unsafe { offsets_l.get_mut().as_mut_ptr() };
+            end_l = unsafe { offsets_l.get_mut().as_mut_ptr() };
```
Isn't this also invalid? I believe offsets_l is uninitialized at this point?

Mark-Simulacrum commented on src/libcore/slice/sort.rs:
```diff
@@ -288,8 +288,8 @@ fn partition_in_blocks<T, F>(v: &mut [T], pivot: &T, is_less: &mut F) -> usize
             // Trace `block_r` elements from the right side.
-            start_r = offsets_r.as_mut_ptr();
-            end_r = offsets_r.as_mut_ptr();
+            start_r = unsafe { offsets_r.get_mut().as_mut_ptr() };
```
This code also looks like this should be invalid? At least, `offsets_r` is not necessarily initialized, right?

scottmcm commented on src/liballoc/collections/btree/node.rs:
```diff
@@ -83,8 +83,8 @@ struct LeafNode<K, V> {
-    keys: [K; CAPACITY],
-    vals: [V; CAPACITY],
+    keys: MaybeUninit<[K; CAPACITY]>,
+    vals: MaybeUninit<[V; CAPACITY]>,
```
Based on the comment above, should this instead be `[MaybeUninit<V>; CAPACITY]`?

(That might be a pain, though, and this is no worse than before, so might not be this PR.)

RalfJung replied:
That seems like a pretty equivalent type to me. Which comment above?

eddyb replied:
Should we document that you can partially-initialize `MaybeUninit`, i.e. that it "distributes over products"?

RalfJung replied:
I guess this is forth saying, yes.

Unfortunately actually implementing this is not easy, same problem as for `Cell` (which also distributes over products).

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/418000493) ([raw log](https://api.travis-ci.org/v3/job/418000493/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:52:51] ....................................................................................................
[00:52:54] ....................................................................................................
[00:52:57] .....i..............................................................................................
[00:53:00] ....................................................................................................
[00:53:03] ......................................................iiiiiiiii.....................................
[00:53:08] ....................................................................................................
[00:53:12] ....................................................................................................
[00:53:15] ...................................i................................................................
[00:53:17] .....................................................................................i.i..ii........
---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-gdb (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:01:01] 
[01:01:01] running 110 tests
[01:01:12] iiii.......i..i........i..i.i.............i..........iiii...........i.F..i....F.....ii.i.i.......ii.
[01:01:13] failures:
[01:01:13] 
[01:01:13] ---- [debuginfo-gdb] debuginfo/nil-enum.rs stdout ----
[01:01:13] ---- [debuginfo-gdb] debuginfo/nil-enum.rs stdout ----
[01:01:13] NOTE: compiletest thinks it is using GDB without native rust support
[01:01:13] NOTE: compiletest thinks it is using GDB version 7011001
[01:01:13] 
[01:01:13] error: line not found in debugger output: $1 = {<No data fields>}
[01:01:13] status: exit code: 0
[01:01:13] command: "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/nil-enum/nil-enum.debugger.script"
[01:01:13] ------------------------------------------
[01:01:13] GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
[01:01:13] Copyright (C) 2016 Free Software Foundation, Inc.
[01:01:13] Copyright (C) 2016 Free Software Foundation, Inc.
[01:01:13] License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
[01:01:13] This is free software: you are free to change and redistribute it.
[01:01:13] There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
[01:01:13] and "show warranty" for details.
[01:01:13] This GDB was configured as "x86_64-linux-gnu".
[01:01:13] Type "show configuration" for configuration details.
[01:01:13] For bug reporting instructions, please see:
[01:01:13] <http://www.gnu.org/software/gdb/bugs/>.
[01:01:13] Find the GDB manual and other documentation resources online at:
[01:01:13] <http://www.gnu.org/software/gdb/documentation/>.
[01:01:13] For help, type "help".
[01:01:13] Type "apropos word" to search for commands related to "word".
[01retty-std-collections.debugger.script"
[01:01:13] ------------------------------------------
[01:01:13] GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
[01:01:13] Copyright (C) 2016 Free Software Foundation, Inc.
[01:01:13] Copyright (C) 2016 Free Software Foundation, Inc.
[01:01:13] License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
[01:01:13] This is free software: you are free to change and redistribute it.
[01:01:13] There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
[01:01:13] and "show warranty" for details.
[01:01:13] This GDB was configured as "x86_64-linux-gnu".
[01:01:13] Type "show configuration" for configuration details.
[01:01:13] For bug reporting instructions, please see:
[01:01:13] <http://www.gnu.org/software/gdb/bugs/>.
[01:01:13] Find the GDB manual and other documentation resources online at:
[01:01:13] <http://www.gnu.org/software/gdb/documentation/>.
[01:01:13] For help, type "help".
[01:01:13] Type "apropos word" to search for commands related to "word".
[01:01:13] Breakpoint 1 at 0x144f6: file /checkout/src/test/debuginfo/pretty-std-collections.rs, line 57.
[01:01:13] [Thread debugging using libthread_db enabled]
[01:01:13] Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[01:01:13] 
[01:01:13] Breakpoint 1, pretty_std_collections::main::h617474859dc5b54e () at /checkout/src/test/debuginfo/pretty-std-collections.rs:57
[01:01:13] 57     zzz(); // #break
[01:01:13] $1 = BTreeSet<i32>(len: 3)
[01:01:13] $2 = BTreeMap<i32, i32>(len: 3)
[01:01:13] $3 = VecDeque<i32>(len: 3, cap: 8) = {5, 3, 7}
[01:01:13] A debugging session is active.
[01:01:13] 
[01:01:13]  Inferior 1 [process 15246] will be killed.
[01:01:13] 
[01:01:13] Quit anyway? (y or n) [answered Y; input not from terminal]
[01:01:13] ------------------------------------------
[01:01:13] stderr:
[01:01:13] ------------------------------------------
[01:01:13] ------------------------------------------
[01:01:13] Python Exception <class 'gdb.error'> Cannot subscript requested type.: 
[01:01:13] Python Exception <class 'gdb.error'> Cannot subscript requested type.: 
[01:01:13] ------------------------------------------
[01:01:13] 
[01:01:13] thread '[debuginfo-gdb] debuginfo/pretty-std-collections.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3189:9
[01:01:13] 
---
[01:01:13] 
[01:01:13] thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:497:22
[01:01:13] 
[01:01:13] 
[01:01:13] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/debuginfo" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "debuginfo-gdb" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-5.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "5.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:01:13] 
[01:01:13] 
[01:01:13] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:01:13] Build completed unsuccessfully in 0:12:31
[01:01:13] Build completed unsuccessfully in 0:12:31
[01:01:13] make: *** [check] Error 1
[01:01:13] Makefile:58: recipe for target 'check' failed

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0ef71254
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
The command "date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
" exited with 0.
travis_fold:start:after_failure.1
travis_time:start:0aaf1540
$ echo/bootstrap-1v3ifugz4t07z/s-f40eeanft4-1h7wppw-1iplbejydu42a
128740 ./obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release
128648 ./obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu
128644 ./obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release
125880 ./obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

eddyb commented on src/librustc_target/abi/mod.rs:
```diff
@@ -784,6 +784,14 @@ impl Abi {
     }
+
+    /// Returns true if this is an uninhabited type
+    pub fn is_uninhabited(&self) -> bool {
```
For some reason I would've avoided this but I guess it's fine.
Could also go through `rg '==.*::Uninhabited'` and change all of those to use `is_uninhabited`?

eddyb commented on src/librustc_codegen_llvm/mir/block.rs:
```diff
@@ -462,6 +462,55 @@ impl FunctionCx<'a, 'll, 'tcx> {
+                    } else {
+                        "Attempted to instantiate an uninhabited type (e.g. `!`) \
+                         using mem::uninitialized()"
+                    };
```
You can actually use `format!("... uninhabited type `{}` using ...", sig.output())` here, heh.

eddyb commented on src/librustc_codegen_llvm/mir/block.rs:
```diff
@@ -462,6 +462,55 @@ impl FunctionCx<'a, 'll, 'tcx> {
                 }
 
+                if (intrinsic == Some("init") || intrinsic == Some("uninit")) &&
+                    bx.cx.layout_of(sig.output()).abi.is_uninhabited()
```
Can you move this below `let fn_ty`? Then you can use `fn_ty.ret.layout` instead of calling `layout_of`.

eddyb commented on src/test/run-pass/panic-uninitialized-zeroed.rs:
```diff
@@ -0,0 +1,31 @@
+        panic::catch_unwind(|| mem::zeroed::<!>()).is_err();
+
+        panic::catch_unwind(|| mem::uninitialized::<Foo>()).is_err();
+        panic::catch_unwind(|| mem::zeroed::<Foo>()).is_err();
```
I think you can also do `assert_eq!(....downcast_ref::<&'static str>(), Some("... the panic message ..."))` here.

eddyb commented:
Looks like the gdb pretty-printing scripts make assumptions about the fields of libstd datastructures:
https://github.com/rust-lang/rust/blob/6bf6d50a6ff7685b4aa09172d9d09f03f250da9d/src/etc/debugger_pretty_printers_common.py#L392-L401

RalfJung commented on src/liballoc/collections/btree/node.rs:
```diff
@@ -567,7 +567,7 @@ impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Immut<'a>, K, V, Type> {
             unsafe {
                 slice::from_raw_parts(
-                    self.as_leaf().keys.as_ptr(),
+                    self.as_leaf().keys.get_ref().as_ptr(),
```
Here and below, can't you use `MaybeUninit::as_ptr`/`MaybeUninit::as_mut_ptr`?

RalfJung commented on src/libcore/fmt/float.rs:
```diff
@@ -20,11 +20,11 @@ fn float_to_decimal_common_exact<T>(fmt: &mut Formatter, num: &T,
         let formatted = flt2dec::to_exact_fixed_str(flt2dec::strategy::grisu::format_exact,
                                                     *num, sign, precision,
-                                                    false, &mut buf, &mut parts);
+                                                    false, buf.get_mut(), parts.get_mut());
```
Another case of using `&mut` as `&out`.

RalfJung commented on src/libcore/mem.rs:
```diff
@@ -1025,6 +1027,100 @@ impl<T: ?Sized> DerefMut for ManuallyDrop<T> {
+    /// Note that dropping a `MaybeUninit` will never call `T`'s drop code.
+    /// It is your responsibility to make sure `T` gets dropped if it got initialized.
+    #[unstable(feature = "maybe_uninit", issue = "53491")]
+    pub const fn uninitialized() -> MaybeUninit<T> {
```
Why not stabilize as `const fn`?

RalfJung commented on src/test/run-pass/panic-uninitialized-zeroed.rs:
```diff
@@ -0,0 +1,31 @@
+struct Foo {
+    x: u8,
+    y: !,
+}
```
Could you add a test for an empty enum?

Also, AFAIK `Foo` is NOT layout-uninhabited, because of https://github.com/rust-lang/rust/issues/49298?

eddyb replied:
Correct, you need to change `x`'s type to `()`.

eddyb replied:
Wait, no, check this out:
```rust
#![feature(never_type)]
pub struct Foo { x: u8, y: ! }
pub fn foo() -> Foo { loop {} }
```
That gets `noreturn` in the LLVM IR, with `u8` or `()` in the first field, but if you change the `!` to `()`, you lose the `noreturn`.

What happened in the first for #49298 is we *kept* the layout-uninhabited property *propagated*, but it wouldn't affect type sizes the same.

Which means we can use the check here to detect `noreturn` UB in all relevant cases!

RalfJung replied:
`nocall` would be more correct than `noreturn` though :P

Which "check here"?

eddyb replied:
I mean the panicking intrinsics.

bors commented:
:umbrella: The latest upstream changes (presumably #53530) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented:
I finally [wrote that blog post](https://internals.rust-lang.org/t/two-kinds-of-invariants/8264) about whether and when `&mut T` must be initialized.

pietroalbini commented:
Ping from triage @RalfJung! This PR needs your review.

bors commented:
:umbrella: The latest upstream changes (presumably #53227) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented:
There are still several uses of `&mut` as `&out` in `src/libcore/fmt/float.rs` -- if the rules in the `MaybeUninit` doc are strictly applied, that could would be UB.  However, it is certainly no more UB than it was before this PR, so I am included to accept anyway.  What would be interesting to know (as a data point for the unsafe code guidelines around `&mut`) is how hard it would be to change that code to use stretch the (current) rules. @japaric can you have a look?

Otherwise, r=me once Travis is happy with the gdb stuff.

japaric commented:
Sorry for the delay! I'm looking again at this. This is one of the debuginfo tests that's failing:

https://github.com/rust-lang/rust/blob/fea32f1b775b3f37fc4abfa6391c1bebe48af9d1/src/test/debuginfo/nil-enum.rs#L30-L43

The test is instantiating nil enums and then inspecting them in GDB. The program now panics with the changes mode in this PR so the test broke.

I'm not quite sure what to do here. Should I instantiate `MaybeUninit<ANilEnum>` and inspect that? (Would that make the test moot since it would become a union test?). Should I continue to instantiate `ANilEnum` using `MaybeUninit::zeroed().into_inner()`? (Sounds bad because UB) Or should I just remove the test? (because no one should be instantiating nil enums anyway).

cc @michaelwoerister

RalfJung commented:
Yeah that test is definitely UB. It might still be worth seeing how gbd displays that.

So maybe add a comment saying it is UB, and then use `MaybeUninit::uninitialize().into_inner()`? That replicates the previous code as closely as possible.

japaric commented:
Rebased and fixed the debuginfo tests in the last two commits. Regarding using `MaybeUninit` in `fmt/float.rs` I see two ways to replace `&mut [T]` in argument position where the passed buffer is uninitialized:

- Use [`&mut MaybeUninit<[T; N]>`](https://github.com/japaric/rust/commit/7947edb96d43798c64dec0767881989aa0585d84)
- Use [`&mut [MaybeUninit<T>]`](https://github.com/japaric/rust/commit/f82f7f0b2155aab27fad8f692bb572e83f731e30)

I have explored the two options by re-implementing one function using both approaches. I have pushed my work in two branches which are linked above.

Neither feels ergonomic (the first approach actually feels wrong because it constrains the argument) but the second option is slightly better except for the fact that instantiating `[MaybeUninit<T>; N]` requires writing `MaybeUninit::uninitialized()` N times. Could we make the compiler accept `[MaybeUninit<T>; N]`?

abonander commented:
Is it UB to transmute `MaybeUninit<[T; N]>` to `[MaybeUninit<T>; N]`? That'd be a more convenient way to get the latter.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/423443529) ([raw log](https://api.travis-ci.org/v3/job/423443529/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain

[00:04:37] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:04:37] tidy error: /checkout/src/etc/gdb_rust_pretty_printing.py:325: line longer than 100 chars
[00:04:37] tidy error: /checkout/src/etc/gdb_rust_pretty_printing.py:350: line longer than 100 chars
[00:04:37] tidy error: /checkout/src/etc/gdb_rust_pretty_printing.py:354: line longer than 100 chars
[00:04:39] some tidy checks failed
[00:04:39] 
[00:04:39] 
[00:04:39] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:39] 
[00:04:39] 
[00:04:39] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:39] Build completed unsuccessfully in 0:00:49
[00:04:39] Build completed unsuccessfully in 0:00:49
[00:04:39] Makefile:79: recipe for target 'tidy' failed
[00:04:39] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:1cad2c55
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:1534a07f:start=1535830235551357449,finish=1535830235559531005,duration=8173556
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:00227130
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:1b235580
travis_time:start:1b235580
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:029afa78
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

japaric commented:
@abonander I think it's not UB but (last time I checked) transmute doesn't work in const context so you can't use that approach to create a `static mut X: [MaybeUninit<u8>; 1024]`. Also using transmute in non-const context sounds more error prone than convenient to me.

@RalfJung I have fixed the debuginfo issues and CI is green. Do you want me to do anything else before landing this PR?

BatmanAoD commented:
Since this is assigned to @RalfJung and @japaric has addressed the existing comments, I'm changing the tag to waiting-on-review.

RalfJung commented:
I'm on vacation, will be back in a week.

abonander commented:
I didn't think of this before to suggest it on the RFC and I wasn't able to follow the discussion to see if this was already pitched, but if we can't guarantee that `Box::new(MaybeUninit::uninitialized())` *won't* copy a bunch of garbage data from the stack, can we provide a constructor that can? It could probably just wrap `box MaybeUninit::uninitialized()`.

tromey commented:
> Yeah that test is definitely UB. It might still be worth seeing how gbd displays that.

Note that with the enum debuginfo changes (#54004), this particular test provokes a gdb crash.  See the [gdb bug](https://sourceware.org/bugzilla/show_bug.cgi?id=23626).  I have a fix for this, but if this test is in fact invalid, I would like to know sooner rather than later, since then I could just drop this patch.

eddyb commented:
@tromey The test is UB, and rustc and LLVM are allowed to remove all code after the creation of a value of an `enum Foo {}` type (i.e. it's "unreachable" to create such a value).

But gdb shouldn't crash, and you could test gdb on a more appropriate example, involving e.g. pattern-matching on `Result<T, Void>` where `enum Void {}`.

RalfJung commented:
@abonander It definitely shouldn't copy. @japaric can you add a codegen test to this effect? Looking at https://play.rust-lang.org/?gist=e80eed1cd66867d72bc7680be752818d&version=stable&mode=release&edition=2015, it should probably make sure that there is no `store` and no `memset` in the IR.

@eddyb ah so the test is about just having an empty enum "around", not about having a value of it?

RalfJung commented:
I think for the floating point stuff, it is fine to accept the PR as-is, and come back to this once we decided about validity of references. Your work with these two implementations will certainly be a helpful datapoint in that discussion, @japaric -- thanks!

So, as far as this PR is concerned, I think it is ready modulo the aforementioned codegen test. I am fine with the gdb tests but I also have no idea what they are even checking, so I'll leave it to @eddyb whether we can accept as-is or should do further changes here.

abonander commented:
@RalfJung did you mean `memcpy()`? I don't think that will be enough because it appears LLVM has a high threshold for switching from emitting `mov`s to a `memcpy()`.

I think we could instead do a runtime test that writes some arbitrary bytes to the stack above the stack pointer and sees if they're returned in the value or if it's zeroed (as it should be from a fresh allocation).

RalfJung commented:
@abonander yeah that's what I meant.

japaric commented:
I added a codegen test.

@RalfJung does it look like what you had in mind?

RalfJung commented:
@japaric looks good!

RalfJung commented:
@bors r+

bors commented:
:pushpin: Commit d452703fcd9bfd7e6e8771f8be9de88446d0828e has been approved by `RalfJung`

<!-- @bors r=RalfJung d452703fcd9bfd7e6e8771f8be9de88446d0828e -->

bors commented:
:hourglass: Testing commit d452703fcd9bfd7e6e8771f8be9de88446d0828e with merge beb1f045e43db7949b6113fb3ece415996e64cc9...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/431520318?utm_source=github_status&utm_medium=notification)

rust-highfive commented:
The job `wasm32-unknown` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/431520355) ([raw log](https://api.travis-ci.org/v3/job/431520355/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[01:14:50] test [run-pass] run-pass/wrong-hashset-issue-42918.rs ... ok
[01:14:50] 
[01:14:50] failures:
[01:14:50] 
[01:14:50] ---- [run-pass] run-pass/panic-uninitialized-zeroed.rs stdout ----
[01:14:50] error: test run failed!
[01:14:50] status: exit code: 101
[01:14:50] status: exit code: 101
[01:14:50] command: "/node-v9.2.0-linux-x64/bin/node" "/checkout/src/etc/wasm32-shim.js" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/panic-uninitialized-zeroed/a.wasm"
[01:14:50] ------------------------------------------
[01:14:50] 
[01:14:50] ------------------------------------------
[01:14:50] stderr:
[01:14:50] stderr:
[01:14:50] ------------------------------------------
[01:14:50] RuntimeError: unreachable
[01:14:50]     at __rust_start_panic (wasm-function[101]:1)
[01:14:50]     at rust_panic.llvm.15713727278457583197 (wasm-function[95]:38)
[01:14:50]     at std::panicking::rust_panic_with_hook::h226a58a5977f2e17 (wasm-function[90]:500)
[01:14:50]     at std::panicking::continue_panic_fmt::h5f0126bd5d9da93f (wasm-function[89]:155)
[01:14:50]     at rust_begin_unwind (wasm-function[88]:3)
[01:14:50]     at core::panicking::panic_fmt::hebc42c0af7746402 (wasm-function[185]:80)
[01:14:50]     at core::panicking::panic::hdebf053aeca5c9b4 (wasm-function[184]:109)
[01:14:50]     at core::mem::uninitialized::hc5b2f0d7a16034e3 (wasm-function[24]:7)
[01:14:50]     at panic_uninitialized_zeroed::main::_$u7b$$u7b$closure$u7d$$u7d$::h25c1ee37bf2b1209 (wasm-function[19]:1)
[01:14:50]     at std::panicking::try::do_call::hb35f4db8c63cce7c (.llvm.10049803345321850598) (wasm-function[13]:1)
[01:14:50]     at __rust_maybe_catch_panic (wasm-function[100]:5)
[01:14:50]     at panic_uninitialized_zeroed::main::h3462c8ec1e7650c0 (wasm-function[8]:60)
[01:14:50]     at std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::h6a977ae82a4b77c9 (wasm-function[4]:25)
[01:14:50]     at std::sys_common::backtrace::__rust_begin_short_backtrace::h178ce1cdbcfd3352 (wasm-function[51]:8)
[01:14:50]     at std::panicking::try::do_call::h92ed3f8ab1f1fe4c (.llvm.15713727278457583197) (wasm-function[87]:20)
[01:14:50]     at __rust_maybe_catch_panic (wasm-function[100]:5)
[01:14:50]     at std::rt::lang_start_internal::had2eaac6c6c88dfd (wasm-function[52]:128)
[01:14:50]     at std::rt::lang_start::h7233662e55760c3a (wasm-function[3]:42)
[01:14:50]     at main (wasm-function[9]:11)
[01:14:50]     at Object.<anonymous> (/checkout/src/etc/wasm32-shim.js:136:20)
[01:14:50] ------------------------------------------
[01:14:50] 
[01:14:50] 
[01:14:50] thread '[run-pass] run-pass/panic-uninitialized-zeroed.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:3205:9
[01:14:50] 
[01:14:50] 
[01:14:50] failures:
[01:14:50]     [run-pass] run-pass/panic-uninitialized-zeroed.rs
[01:14:50]     [run-pass] run-pass/panic-uninitialized-zeroed.rs
[01:14:50] 
[01:14:50] test result: FAILED. 481 passed; 1 failed; 68 ignored; 0 measured; 0 filtered out
[01:14:50] 
[01:14:50] 
[01:14:50] 
[01:14:50] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/wasm32-unknown-unknown/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/run-pass" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass" "--stage-id" "stage2-wasm32-unknown-unknown" "--mode" "run-pass" "--target" "wasm32-unknown-unknown" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--nodejs" "/node-v9.2.0-linux-x64/bin/node" "--host-rustcflags" "-Crpath -O -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/wasm32-unknown-unknown/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--llvm-version" "8.0.0svn\n" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:14:50] 
[01:14:50] 
[01:14:50] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --target wasm32-unknown-unknown src/test/run-make src/test/ui src/test/run-pass src/test/compile-fail src/test/parse-fail src/test/mir-opt src/test/codegen-units src/libcore
[01:14:50] Build completed unsuccessfully in 1:11:11
---
travis_time:end:0241a3b2:start=1537544182071536243,finish=1537544182075900478,duration=4364235
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0461f89a
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:02fa98a1
travis_time:start:02fa98a1
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:0c760937
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

RalfJung commented:
Seems something doesn't work right on wasm?

RalfJung commented:
Seems fine to me. I'll drop a note in chat so WASM people can complain if that's wrong.

japaric commented:
added `ignore-wasm32-bare`

@bors r=RalfJung

bors commented:
:pushpin: Commit 42972c3c9e7a8ba6d6e3ec906fb5aa6be86e76b3 has been approved by `RalfJung`

<!-- @bors r=RalfJung 42972c3c9e7a8ba6d6e3ec906fb5aa6be86e76b3 -->

ashleygwilliams commented:
hey folks- realized you are running into something that the rust wasm working group has a tool for- https://github.com/rustwasm/console_error_panic_hook. this tool will log panics with `console.error`. i haven' t dug fully into your tests but this might make this situation more pleasant.

japaric commented:
@ashleygwilliams Thanks for the input!

I believe these run-pass tests don't provide an easy way to depend on crates from crates.io but we could write a run-make test that makes use of the console-error-panic-hook crate to check that the panic message printed to the console matches the expected output. 

@RalfJung do you want me to do that ^ in this PR? Or should we leave it up for a follow-up PR? It sounds like quite a bit of work, tbh. If we want to add some wasm test in this PR it might be easier to write some sort of run-fail test that checks that `mem::uninitialized` produces an `unreachable` error (as seen in the log of the failed Travis CI build) and we omit checking the panic message (as there's none) in that test.

Removing from the bors queue until a decision is made.

@bors r-

alexcrichton commented:
Oh this test is correct to have `// ignore-wasm32-bare`, panics aren't implemented and this test relies on catching panics. This isn't related to getting a good panic message or debugging what's going on.

@bors: r=RalfJung

bors commented:
:pushpin: Commit 42972c3c9e7a8ba6d6e3ec906fb5aa6be86e76b3 has been approved by `RalfJung`

<!-- @bors r=RalfJung 42972c3c9e7a8ba6d6e3ec906fb5aa6be86e76b3 -->

bors commented:
:hourglass: Testing commit 42972c3c9e7a8ba6d6e3ec906fb5aa6be86e76b3 with merge aacd85525d1ced618b4bf896e0835140a996f85f...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.9073)

RalfJung commented:
```
---- [run-pass] run-pass\panic-uninitialized-zeroed.rs stdout ----
normalized stderr:
warning: field is never used: `x`
  --> $DIR/panic-uninitialized-zeroed.rs:20:5
   |
LL |     x: u8,
   |     ^^^^^
   |
   = note: #[warn(dead_code)] on by default
warning: field is never used: `y`
  --> $DIR/panic-uninitialized-zeroed.rs:21:5
   |
LL |     y: !,
   |     ^^^^
```
Looks like it needs `allow(unused)`

SimonSapin commented:
> Does not deprecate mem::{uninitialized,zeroed} just yet. As per #53491 (comment), we should not deprecate them until MaybeUninit is stabilized.

Note that you can now have a `rustc_deprecated` attribute with a *future* version number in the `since` parameter, and rustc will only emit warnings in that and later versions.

https://github.com/rust-lang/rust/blob/ff7967056916f68f2df5214e47e9e72414b44fc3//src/libcore/str/mod.rs#L3893

japaric commented:
@bors r=RalfJung

bors commented:
:pushpin: Commit 1cdbad20224a2e8da4a63aea8389b34bf8313281 has been approved by `RalfJung`

<!-- @bors r=RalfJung 1cdbad20224a2e8da4a63aea8389b34bf8313281 -->

japaric commented:
@SimonSapin yes, I'm using that but wrote '2.0.0' in the "since" field because I don't have an estimate for when this will be stabilized.

bors commented:
:hourglass: Testing commit 1cdbad20224a2e8da4a63aea8389b34bf8313281 with merge c6e3d7fa3113aaa64602507f39d4627c427742ff...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.9076), [status-travis](https://travis-ci.org/rust-lang/rust/builds/431998360?utm_source=github_status&utm_medium=notification)
Approved by: RalfJung
Pushing c6e3d7fa3113aaa64602507f39d4627c427742ff to master...

Valinora commented:
Was this expected to have a perf impact?
[perf link](https://perf.rust-lang.org/compare.html?start=4591a245c7eec9f70d668982b1383cd2a6854af5&end=c6e3d7fa3113aaa64602507f39d4627c427742ff&stat=instructions:u)

nnethercote commented:
It's a significant perf regression:
```
tokio-webpush-simple-opt
	avg: 14.3%	min: 1.5%	max: 37.6%
issue-46449-opt
	avg: 8.4%	min: 1.3%	max: 13.1%
script-servo-opt
	avg: 4.1%	min: 0.8%	max: 6.4%
issue-46449-debug
	avg: 4.1%	min: 1.0%	max: 4.8%
cargo-opt
	avg: 3.5%	min: 1.4%	max: 4.7%
webrender-opt
	avg: 2.4%	min: 1.1%	max: 3.5%
ripgrep-opt
	avg: 1.8%	min: 1.0%	max: 3.4%
futures-opt
	avg: 1.8%	min: 0.4%	max: 3.4%
script-servo-debug
	avg: 1.8%	min: 1.1%	max: 3.3%
sentry-cli-opt
	avg: 2.3%	min: 1.0%	max: 3.0%
regex-opt
	avg: 1.7%	min: 0.5%	max: 2.9%
tokio-webpush-simple-debug
	avg: 2.3%	min: 2.0%	max: 2.8%
inflate-opt
	avg: -0.9%?	min: -2.3%?	max: 0.7%?
clap-rs-debug
	avg: 0.9%	min: 0.5%	max: 2.1%
ctfe-stress-opt
	avg: -0.8%?	min: -2.1%?	max: 0.5%?
hyper-opt
	avg: 1.2%	min: 0.5%	max: 2.1%
ripgrep-debug
	avg: 1.7%	min: 1.2%	max: 2.0%
futures-debug
	avg: 1.1%	min: 0.4%	max: 2.0%
crates.io-opt
	avg: 1.2%	min: 0.5%	max: 1.9%
html5ever-opt
	avg: 0.6%	min: -0.1%	max: 1.8%
sentry-cli-debug
	avg: 1.2%	min: 0.8%	max: 1.8%
clap-rs-check
	avg: 1.0%	min: 0.6%	max: 1.7%
deeply-nested-check
	avg: 1.2%	min: 0.0%	max: 1.7%
piston-image-opt
	avg: 1.2%	min: 1.0%	max: 1.7%
serde-debug
	avg: 1.1%	min: 0.4%	max: 1.7%
serde-check
	avg: 1.2%	min: 0.2%	max: 1.6%
serde-opt
	avg: 1.0%	min: 0.3%	max: 1.6%
encoding-opt
	avg: 1.0%	min: 0.3%	max: 1.6%
```
@japaric: can you investigate? Here are [docs](https://github.com/rust-lang-nursery/rustc-perf/blob/master/collector/README.md) on benchmarking and profiling. Cachegrind can be particularly useful for examining the difference between two compiler versions.

nnethercote commented:
On second thought, this regression is bad enough that I think we should consider backing the changes out.

nnethercote commented:
@rust-lang/wg-compiler-performance

Mark-Simulacrum commented:
I support backing this change out to investigate performance losses.

RalfJung commented:
Fine for me. I assume the regression comes from one of the places where `mem::uninitialized` was replaced by `MaybeUninit`, so we could then land the new union separately from the part which starts to make use of it in the compiler.

RalfJung commented:
Revert submitted as https://github.com/rust-lang/rust/pull/54554
